### DISCUSSION TITLE
我为什么放弃了layui

### DISCUSSION THREAD

**Discussion by mouse0w0** [layui/layui#2404] [讨论]
~~前端开发偶遇layui，wontfix强如怪物，拼尽全力无法战胜（并非全力）。~~

初次体验了layui两三个月，感觉还是对这个前端框架充满遗憾吧。本人只是一个初入JavaScript的前端，也只会写一些CURD的后端，仅从我个人经验出发，谈谈以下几点：

1. 参数定义的不规范，初次使用layui便已感受到。每个模块的同样功能在参数命名和可接受的参数类型的定义上就各有不同，可以说是每个模块的都有自己的想法，而非整个框架都有一个统一的规范。这点我已在讨论 #2227 提到过。
2. “极简丰盈”的虚假宣传，贫乏的组件和功能令人心累。相较于现在市面上先进的前端UI框架，layui实在是远远落后太多。
3. “返璞归真”只是抱残守缺的借口。现代IDE同样为前端开发提供自动补全、语义检查、实时文档等功能，任何一个后端开发者都可以在前端开发时，感受到如同后端开发一般的丝滑开发体验。而使用layui时，我不得不每行代码都检查文档，以确认每个组件，每个方法，每个参数接受的数值。

以上只是一些我对layui的主观使用感受，并非对layui维护者和贡献者们劳动成果的否定，也并非无故攻击批判layui。layui确实是一个在中文互联网上留下了一笔的框架，但不得不说，或许layui在这个时代真该死了。或就如之前在layui的某个破防老哥的issue里看到，社区某些拥趸所骂：“技术选型错了还怪框架，傻逼”，或许真是我不该遇见layui，不该选中layui，不该成为那个傻逼。

希望layui 3能够真正践行文档所说的目标，我们有缘江湖再见。

**Comment by sentsim** [comment]
看得出来，您是一位不可多得的质量用户，但很遗憾 Layui 并没有为你带来很好的体验。

事实上，**wontfix** 是维护者的一道防线。当然，也可能是你的最后一道防线。因此，当我尝试在两个不同的视觉中切换，我对你的这段文字既有反驳，也有认同。认同的是，对一个工具从初识到毅然放弃的心里转变，是作为开发人员在技术选型时的常有经历，也正因为如此，开发者才能不断优化个人偏好，从而找出更适合自己的技术方案，这何尝不是一种历练与成长呢。遗憾的是，你其中提到的「极简丰盈的虚假宣传」的观点有失理性，更多可能只是情绪上的宣泄。Layui 的部分组件相比之下确实显得简陋，但丰盈且具有一定特点的组件也并非没有，只是将一句「贫乏的组件和功能」砸在这个数年来被广泛应用的传统组件库上，让人有种英雄迟暮的悲凉啊！我不知道什么样的组件才是你认为的丰盈，但至少我在 Layui 的组件代码中交出了我认为的答案。

返璞归真也好，抱残守阙也罢，不过只是一种存在的立场表达。你认可它的存在，它即是返璞归真，反之，就是抱残守阙。Layui 社区持续不断的热情仍然支撑着它的存在，作为维护者，我一定要面向它的受众大声地告诉他们：Layui 是一件多么落后不堪的工具吗？不，没有意义，我们努力维持着它仅存的光与热，即是维持它存在的合理性，维持一种温度、一种希望。

但愿那些先进的 UI 库能让你运用自如。加油少年！

**Comment by hulang** [comment]
组件确实太少了，官方的`第3方`组件，是多，但是有的组件可能还是`3年以前`的，随着layui不断更新，第3方组件`兼容性`不行了。
我多次提到，可以参考一下layui-vue，以：`闲心`老大和`晨星`大佬的功底，多扩展一下成官方组件应该没多大问题。。

**Reply by bxjt123** [reply]
真的，相比`layui-vue`来说，不懂他那边为什么能多出来那么多组件，而且还很实用。
很多次都想换成`vue`，奈何项目太多了，全换了费劲。

**Reply by sentsim** [reply]
Layui 就算再增加 10+ 个组件，一段时间没新增后，还是会有很多人觉得少。
我认为 Layui 当前缺少的不是组件，而是信心。

**Reply by lanrenbulan** [reply]
layui的坚持值得敬重。国家鼓励生育，但也会建全养老。项目的维护也是迭代升级，如果不停地大版本更新，让使用的他的人何其难。layui的存在维护了曾经是身处现代的人。那多年以后，每当回首，曾经最美。

假如你现在不喜欢跳广场舞，那当变成老头时，你也喜欢跟着大妈一起扭。

与其说“我为什么放弃了layui”，不如分享下在使用layui中踩的坑或选择建议，这样会显得更加务实些。

PS：v2.9.21啥时release

**Reply by HuangLei0x2e5** [reply]
其实我觉得组件倒还够用，反正我是做后台管理系统，再复杂的UI我就直接外包，反正现在满街都是用VUE的前端。

我觉得layui的问题，就是太封闭了，说是JQuery吧，又不像JQuery我直接可以改一个参数，太多的东西有自己的一套逻辑了。

**Reply by sentsim** [reply]
@lanrenbulan 元旦后，周六之前发 2.9.21 正式版，并作为 2.x 阶段性稳定版。

**Comment by KingBes** [comment]
首先规范定义是人为的，目前js的规范层出不穷，vue一套、svelte一套、后端一套、跨平台又一套，有些还还几套规则，对于layui这种基于jq的框架来说是比较自由的，如果要写出比较质量的代码，还是比较需要基本功的。
“极简丰盈”这个对于有一定创新组件能力的人来说确实是的，可能对于新人来说也确实不是。
“技术选型”这个无非就是一个选择什么样的工具罢了，对于我个人来说，我是一个全栈开发者，我还是会选择layui这种jq的框架，可能对于大型系统来说没有那么性能规模，但是效率确实比新型框架要高很多的！就算你所为的选型正确，但最终还是沦为“傻逼”！！！技术的本身不是为了追求所谓的攀比和新颖，而是给你带来方便和生活！
